### PR TITLE
Flesh out support for timeseries data.

### PR DESCRIPTION
**Pull request by htmlboss**
Partially addresses #673 

tldr: virtual mooring is back (except for bottom depths)!

FVCOM will come at a later date since @geoffholden has work that depends on the fixes in here, so time is an issue. We don't have any operational FVCOM datasets currently, nor do we expect such datasets in the short-to-medium term.

**Note** timeseries for "bottom" depths won't be addressed in this PR for the above reasoning (i.e. remain broken). I believe there is some numpy axis magic possible to reduce the complexity of our bottom calculations and reduce the volume of code.

## Background

Our timeseries plots have been broken from some time so the basic idea behind these changes is to fix all that. At a lower level, I've tried to simplify the "data getter" functions in `Model.py` so callers can explicitly pass an `endtime` argument to indicate a timeseries is being requested. Before, we were doing implicit checks of the `time`/`timestamp` arguments which resulted in a vague API and code instability.

This PR will re-implement support for timeseries data for any depth level, and "All" depths. "Bottom" depths will come soon.

## Why did you take this approach?
Maybe not the simplest way to do things, but I've opted to delete some dead code (will be commented below) and factor out some code into `Model` thus reducing duplication in the classes deriving from `Model`. This is a first pass at cleaning things up. Further cleaning can come later and is outside the scope of this PR.

## Anything in particular that should be highlighted?
I had to make some minor alterations to our resampling functions in terms of data layout. No changes were made to the actual interpolation though. Screenshots below demonstrate no change in the final result, as intended. Now that we pass slice objects for the time dimension when extracting data, this has caused some issues where the time dim size == 1. To remedy this, I've sprinkled some numpy squeezes to get things back to their expected shape (I compared the shapes of the old resampling vs what I have here and there was a difference in the shapes). Maybe not the "best" approach but the scope was rapidly exploding at that point.

## Screenshot(s)
Virtual mooring (all depths) - Mercator:
image
Virtual mooring (depth 0) - Mercator:
image
Virtual mooring (all depths) - Nemo:
image
Virtual mooring (depth 1) - Nemo:
image

Hovmoller - Nemo:
image
Hovmoller - Mercator:
image

Change comparisons

Profiles:
Before:
image
After:
image

Transects:
Before:
image
After:
image

Deep Sound Channel (calculated):
image

No behaviour change to calculated variables or map tiles:
image

## Checks
- [x] I ran unit tests.
- [x] I've tested the relevant changes from a user POV.

_Hint_ To run all python unit tests run the following command from the root repository directory:
```sh
python -m unittest `find tests -name "*.py" | grep -v disabled`
```

### CODE REVIEW

**Comment by htmlboss** [comment]
Last thing to do is Hovmoller plots

**Comment by htmlboss** [comment]
Type hinting will be coming soon too.

File: data/fvcom.py [hunk 1]
```diff
@@ -247,24 +247,6 @@ def weight(r):
 
         return np.squeeze(output)
 
-    def get_path(self, path, depth, time, variable, numpoints=100, times=None, return_depth=False):
```
Comment (htmlboss) [inline data/fvcom.py]: I've deleted `get_path` from fvcom, nemo, and mercator and moved it into model, since the implementations were identical.

File: data/mercator.py [hunk 2]
```diff
@@ -132,28 +125,17 @@ def __resample(self, lat_in, lon_in, lat_out, lon_out, var, radius=50000):
...
+            output = output.reshape((
+                origshape[0], # time
+                origshape[1], # depth
+                output.shape[0], # lat
+                output.shape[1] # lon
+            ))
 
         return np.squeeze(output)
 
-    def get_path(self, path, depth, time, variable, numpoints=100, times=None, return_depth=False):
```
Comment (htmlboss) [inline data/mercator.py]: I've deleted `get_path` from fvcom, nemo, and mercator and moved it into model, since the implementations were identical.

File: data/nemo.py [hunk 3]
```diff
@@ -177,24 +173,6 @@ def __latlon_vars(self, variable):
 
         raise LookupError("Cannot find latitude & longitude variables")
 
-    def get_path(self, path, depth, time, variable, numpoints=100, times=None, return_depth=False):
```
Comment (htmlboss) [inline data/nemo.py]: I've deleted `get_path` from fvcom, nemo, and mercator and moved it into model, since the implementations were identical.

File: plotting/hovmoller.py [hunk 4]
```diff
@@ -61,50 +59,24 @@ def find_depth(depth, clip_length, dataset):
...
-            if self.starttime > 0:
-                self.starttime = dataset.nc_data.timestamp_to_time_index(
-                    self.starttime)
-            if self.endtime > 0:
-                self.endtime = dataset.nc_data.timestamp_to_time_index(self.endtime)
-
-            time = time_var[slice(min(self.starttime, self.endtime), max(
-                self.starttime, self.endtime) + 1)].values
-
-            if len(self.variables) > 1:
```
Comment (htmlboss) [inline plotting/hovmoller.py]: this if-block is dead code; a relic from when we didn't have the calculation layer to produce the magnitude of two components so we did it manually. There are tons of other spots in the code that do this, but I figured since I'm editing this file already, let's just start cleaning it up.

File: plotting/hovmoller.py [hunk 5]
```diff
@@ -113,39 +85,20 @@ def find_depth(depth, clip_length, dataset):
...
+            with open_dataset(compare_config, timestamp=self.compare['starttime'], endtime=self.compare['endtime'], variable=self.compare['variables']) as dataset:
                 self.compare['depth'], self.compare['depth_value'], self.compare['depth_unit'] = find_depth(
                     self.compare['depth'], len(dataset.depths) - 1, dataset)
 
-                self.fix_startend_times(
-                    dataset, self.compare['starttime'], self.compare['endtime'])
-                time = list(
-                    range(self.compare['starttime'], self.compare['endtime'] + 1))
-
-                if len(self.compare['variables']) > 1:
```
Comment (htmlboss) [inline plotting/hovmoller.py]: dead code for same reason as above.

File: plotting/hovmoller.py [hunk 6]
```diff
@@ -113,39 +85,20 @@ def find_depth(depth, clip_length, dataset):
         # Load data sent from Right Map (if in compare mode)
         if self.compare:
             compare_config = DatasetConfig(self.compare['dataset'])
-            with open_dataset(compare_config, variable=self.compare['variables']) as dataset:
+            with open_dataset(compare_config, timestamp=self.compare['starttime'], endtime=self.compare['endtime'], variable=self.compare['variables']) as dataset:
```
Comment (htmlboss) [inline plotting/hovmoller.py]: This will resolve a crash when trying to open the compared dataset. No other testing done beyond this since it's outside the scope of this PR.

File: plotting/timeseries.py [hunk 7]
```diff
@@ -43,25 +43,24 @@ def parse_query(self, query):
...
-
-            if len(self.variables) == 1:
-                self.variable_unit = self.dataset_config.variable[
-                    dataset.variables[self.variables[0]]
-                ].unit
-                self.variable_name = self.dataset_config.variable[
-                    dataset.variables[self.variables[0]]
-                ].name
-            else:
-                self.variable_name = self.get_vector_variable_name(dataset,
```
Comment (htmlboss) [inline plotting/timeseries.py]: Dead code

File: plotting/timeseries.py [hunk 8]
```diff
@@ -72,36 +71,34 @@ def load_data(self):
...
+            if not (set(dataset.variables[variable].dimensions) &
                     set(dataset.nc_data.depth_dimensions)):
                 self.depth = 0
 
-            times = None
             point_data = []
+            dep = []
             for p in self.points:
                 data = []
-                for v in self.variables:
```
Comment (htmlboss) [inline plotting/timeseries.py]: No need for a loop here since we have the calculation layer to produce correct magnitude values. This plot doesn't support multiple variables on the same plot anyway.

File: plotting/timeseries.py [hunk 9]
```diff
@@ -161,7 +158,7 @@ def csv(self):
                 header.append(["Depth", self.depth])
             else:
                 header.append(
-                    ["Depth", "%d%s" % (np.round(self.depths[self.depth]),
+                    ["Depth", "%.4f%s" % (self.depths,
```
Comment (htmlboss) [inline plotting/timeseries.py]: Increase precision of depth value being saved to CSV.

File: plotting/timeseries.py [hunk 10]
```diff
@@ -132,12 +127,14 @@ def load_data(self):
...
-            self.depths = depths
+            self.depths = dep
             self.depth_unit = "m"
 
+
     def csv(self):
         header = [
-            ['Dataset', self.dataset_name],
+            ['Dataset', self.dataset_config.name],
+            ['Attribution', self.dataset_config.attribution]
```
Comment (htmlboss) [inline plotting/timeseries.py]: Crediting the original dataset is kinda important.

File: plotting/timeseries.py [hunk 11]
```diff
@@ -132,12 +127,14 @@ def load_data(self):
...
             self.data = point_data
-            self.depths = depths
+            self.depths = dep
             self.depth_unit = "m"
 
+
     def csv(self):
         header = [
-            ['Dataset', self.dataset_name],
+            ['Dataset', self.dataset_config.name],
```
Comment (htmlboss) [inline plotting/timeseries.py]: Use the user-friendly dataset name, not the internal snake-case dataset key.

File: plotting/timeseries.py [hunk 12]
```diff
@@ -253,22 +250,12 @@ def csv(self):
         return super(TimeseriesPlotter, self).csv(header, columns, data)
 
     def plot(self):
-        if len(self.variables) > 1:
```
Comment (htmlboss) [inline plotting/timeseries.py]: dead code for same reason as above.

File: plotting/timeseries.py [hunk 13]
```diff
@@ -72,36 +68,34 @@ def load_data(self):
...
 
-            if not (set(dataset.variables[var].dimensions) &
+            # Override depth request if requested variable has no depth (i.e. surface only)
+            if not (set(dataset.variables[variable].dimensions) &
                     set(dataset.nc_data.depth_dimensions)):
                 self.depth = 0
 
-            times = None
             point_data = []
+            dep = []
```
Comment (jamesprayner) [inline plotting/timeseries.py]: I think `dep` should be renamed to `depths` or something similar just for clarity

File: data/calculated_parser/functions.py [hunk 14]
```diff
@@ -80,7 +80,7 @@ def __validate_depth_lat_temp_sal(depth, latitude, temperature, salinity):
     if type(salinity) is not np.ndarray:
         salinity = np.array(salinity)
 
-    return depth, latitude, temperature, salinity
+    return depth, latitude, np.squeeze(temperature), np.squeeze(salinity)
```
Comment (htmlboss) [inline data/calculated_parser/functions.py]: some of that squeezing I was talking about.
